log_cache.go

```diff
@@ -0,0 +1,98 @@
+package raft
+
+import (
+	"sync"
+)
+
+// logCache wraps a logstore with a ring buffer providing fast access to the
+// last n raft log entries.
+type logCache struct {
+	store      LogStore
+	cache      []*Log
+	current    int
+	lastlogidx uint64
+	l          sync.RWMutex
+}
+
+func newLogCache(capacity int, logstore LogStore) *logCache {
+	return &logCache{
+		cache: make([]*Log, 0, capacity),
+		store: logstore,
+	}
+}
+
+func (c *logCache) getLogFromCache(logidx uint64) (*Log, bool) {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	// Return if we have not seen that log entry yet, or
+	// if it was pushed out of the cache already.
+	if logidx > c.lastlogidx || (c.lastlogidx-logidx) >= uint64(len(c.cache)) {
+		return nil, false
+	}
+
+	// 'last' is the index of the element we cached last,
+	// its raft log index is 'lastlogidx'
+	last := (c.current - 1)
+	m := last - int(c.lastlogidx-logidx)
+
+	// See https://golang.org/issue/448 for why (m % n) is not enough.
+	n := cap(c.cache)
+	log := c.cache[((m%n)+n)%n]
```

Comment:
As an example, all of this juggling could be simplified if you can assume length == capacity